src/Core/Aetherling/Passes/ThroughputModifications.hs

```diff
@@ -1,89 +1,42 @@
-increaseLBPxPerClock p img mult | mult == 1 = (p, 1)
--- if not filling out this dimension, pInner * mult must divide into imgInner
--- no need to recurse further as done filling out dimensions
-increaseLBPxPerClock (pInner:pTl) (imgInner:imgTl) mult |
-  imgInner > pInner * mult && (imgInner `mod` (pInner * mult) == 0) =
-  ((pInner * mult) : pTl, mult)
--- if filling out this dimension, imgInner must divide into mult * pInner cleanly
-increaseLBPxPerClock (pInner:pTl) (imgInner:imgTl) mult |
-  imgInner <= pInner * mult && ((mult * pInner) `mod` imgInner == 0) =
-  (imgInner : pOuter, mult * multOuter)
-  where
-    -- since requiring pInner to always divide into imgInner, this is ok
-    remainingMultForOuterDims = (mult * pInner) `ceilDiv` imgInner
-    (pOuter, multOuter) = increaseLBPxPerClock pTl imgTl remainingMultForOuterDims
-increaseLBPxPerClock p _ _ = (p, 1)
-
--- given a LB's pxPerClock, its image dimesions, and a divisor to slow down,
--- speed up the pxPerCLock from inner most to outer most.
--- Returns the new pxPerClock (in reverse order of LB) and the amount sped up
--- NOTE: pxPerClcok and imgDimensions come in same order as for LB
-decreaseLBPxPerClock :: [Int] -> [Int] -> Int -> ([Int], Int)
--- if no more dimensions to speed up, return remaining pxPerClock 
-decreaseLBPxPerClock [] _ _ = ([], 1)
--- if not reducing to 1 out this dimension, pOuter / div must divide into imgOuter
--- and div must divide into pOuter.
--- no need to recurse further as done filling out dimensions
-decreaseLBPxPerClock (pOuter:pTl) (imgOuter:imgTl) div |
-  div < pOuter && (pOuter `mod` div == 0) &&
-  ((imgOuter * div) `mod` pOuter == 0) = ((pOuter `ceilDiv` div) : pTl, div)
--- if reducing dim to 1 px per clock, pOuter must must divide into div  cleanly
-decreaseLBPxPerClock (pOuter:pTl) (imgOuter:imgTl) div |
-  div >= pOuter && (div `mod` pOuter == 0) = 
-  (1 : pInner, div * divInner)
-  where
-    -- since requiring pOuter to always divide into imgOuter, this is ok
-    remainingDivForInnerDims = div `ceilDiv` pOuter
-    (pInner, divInner) = decreaseLBPxPerClock pTl imgTl remainingDivForInnerDims
-decreaseLBPxPerClock p _ _ = (p, 1)
+speedUp throughMult op | actualMult == throughMult = spedUpOp
+  where (spedUpOp, actualMult) = attemptSpeedUp throughMult op
+speedUp throughMult op = Failure
+  (InvalidThroughputModification throughMult actualMult) 
+  where (spedUpOp, actualMult) = attemptSpeedUp throughMult op
 
--- helper function for linebuffer speedUpIfPossible, goes through, increasing
--- parallelism of each component take an op and make it run x times faster
--- without wrapping it, where x is the second argument.
--- This may not be possible for some ops without wrapping them in a map
--- so the pair returns a best effort speed up op and the amount it was sped up
---
 -- If its combinational, pretty much always just wrap it in map, as if later
 -- speed up again, will just increase multiple on map, never see this again
```

Comment:
This comment is cryptic and difficult to understand. Please clarify so others will understand.

In general, I was thinking you'd organize this file in terms of the categories of ops.  In front of each category, you can write a short discussion about what the strategy/issues are for speeding up this category of operations.

Reply:
I now see that some discussion of this is at the top of AST.hs, but it is relevant to this file, not the general AST.  I'd move that comment to the top of this file, and then I'd like to see the four groups clearly delineated by comments in ThroughputModifications.hs.

Reply:
I'd also like a comment documenting some of the assumptions here. When speeding these ops up, you throw a map around the ops since we're assuming that they are really defined on singleton argument types (e.g. int).  

In other words, for simplicity, it's as if there is only ADD T_INT as base operation in AE. If one wanted a ADD T_ARRAY N T_INT, then they's make it by speeding up ADD T_INT by a factor of N, yielding MAP N (ADD T_INT).

Reply:
I'm going to make the improvements we previously discussed (including reorganizing ThroughputModifications.hs) in the next pull request. This is a refactoring of the code base I want in now so that @akeley98 can base his work off it.

Reply:
Oh okay, then don't wait on me to do that. Just push it now.

Reply:
Done. I've been waiting to accept this PR to start the policy of waiting 24 hours before pulling in significant changes to allow time for comment. While this PR is not the ThroughputModifications.hs changes we had discussed, it is a significant change. There may have been discussion on the structure I had chosen that is pertinent to the PR.